planner/optimize.go

```diff
@@ -133,20 +162,33 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 	}
 
 	useBinding := sessVars.UsePlanBaselines
-	stmtNode, ok := node.(ast.StmtNode)
-	if !ok {
+	stmtNode, isStmtNode := node.(ast.StmtNode)
+	if !isStmtNode {
 		useBinding = false
 	}
 	bindRecord, scope, match := matchSQLBinding(sctx, stmtNode)
 	if !match {
 		useBinding = false
 	}
-	if ok {
+	if isStmtNode {
 		// add the extra Limit after matching the bind record
 		stmtNode = core.TryAddExtraLimit(sctx, stmtNode)
 		node = stmtNode
 	}
 
+	// try to get Plan from the General Plan Cache
+	if sctx.GetSessionVars().EnableGeneralPlanCache &&
+		isStmtNode &&
+		!useBinding { // TODO: support binding
+		cachedPlan, names, ok, err := getPlanFromGeneralPlanCache(ctx, sctx, stmtNode, is)
+		if err != nil {
+			return nil, nil, err
+		}
```

Comment:
Should we fall back to the normal path?

Reply:
Actually, I have thought about it, but on second thought, I think throwing the error directly here can help us find more issues when testing.
So I suggest that we let it throw the error directly for now, and then change it to fall-back later after we finish some tests.